### ISSUE TITLE
Request for opinion: adding casing options

### ISSUE THREAD

**Issue by adrienlucbert** [Mateiadrielrafael/scrap.nvim#1]
Greetings,
I'm working on a PR to add casing options to the [`with_casing`](https://github.com/Mateiadrielrafael/scrap.nvim/blob/main/lua/scrap/internal.lua#L442) function, to support some of [`vim-abolish` casing options](https://github.com/tpope/vim-abolish/blob/master/plugin/abolish.vim#L107).
I'd like your opinion @Mateiadrielrafael on how you would like this integrated, as it may introduce breaking changes. Also if you're against adding this to your plugin, feel free to let me know. Here are my propositions:

**Case 1**:
Add casing functions such as `snake_case`, `camel_case`, ..., and add an option for each to `ScrapExpansionOptions`. The API would not necessarily change (unless we change the wording of the current `capitalized` and `all_caps` to something else for the sake of consistency), which would give something like:
```lua
scrap.expand_many(patterns, {capitalized = false, snake_case = true, camel_case = true})
```

**Case 2**:
in order to give more control to the user, I would prefer giving the user the opportunity to provide custom transformer functions that could then be referenced in the `ScrapExpansionOptions` in the call to `expand`. Something like:
```lua
scrap.setup({
    transformers: {
        camel_case = function(word) {
            ...
        },
        snake_case = function(word) {
            ...
        },
    }
})

scrap.expand_many(patterns, {capitalized = false, snake_case = true})
```
This introduces a new `setup` function, but no other breaking changes.

**Comment by Mateiadrielrafael** [comment]
Hi!

So if I understand things correctly, this would let people use the pattern `{ "FooBar", "BarFoo" }` to generate abbreviations for both `FooBar -> BarFoo` and `foo_bar -> bar_foo`. While I don't have a use for this feature, I'd be glad to merge a pr implementing it!

As for your API ideas, I think the second one is probably better, as it enables the user to extend the core however they want. I assume we would also provide some common transformers within the plugin (like snake or camel case). Would these auto-detect the existing casing?

I'd also be open to finding better names for the existing casing options. Although a breaking change, I imagine not many people are currently using this plugin, so I think it's probably fine :thinking:

**Comment by adrienlucbert** [comment]
Hi!
Thanks a lot for considering my proposal. Yes, you understood the goal of it. I find a use for it when trying to implement vim-abolish's :Subvert feature.
How I imagined it, it would work exactly the same way vim-abolish does (matter of fact I would simply port the existing code to lua), converting a word from any case to the desired case. See the code I linked in my comment above, the `snake_case` function for instance ensures a conversion to snake case from any other case, by applying multiple substitutions on top of one another. Hence, no need to auto-detect the current case.
I also agree the second API proposal is a better one, and we could totally imagine providing some common tranformers.
As for the existing ones, I was thinking `all_caps` could be renamed to `uppercase` to match vim-abolish's names, but then vim-abolish's `uppercase` is not the same as yours. It turns to uppercase the snake case variant of the word: `FooBar` -> `FOO_BAR`, when yours does `FooBar` -> `FOOBAR`.
Same goes for `capitalized`, it doesn't match `mixedCase` nor `CamelCase`, I have even fewer ideas how to name it.
I think for now we can leave them at that and see if anything comes to mind later. I'll start working on the PR tomorrow.
Cheers!
